### PR TITLE
fix SQL persistence for materialized view with ASSERT NOT NULL

### PR DESCRIPTION
**Pull request by umanwizard**
### Motivation

  * This PR fixes a recognized bug.

    Discovered by @def- while testing: https://github.com/MaterializeInc/materialize/pull/22339

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

This PR does NOT have adequate test coverage by itself, but this companion PR: https://github.com/MaterializeInc/materialize/pull/22339 should have adequate coverage.

- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Fixes a bug where `ASSERT NOT NULL` options on materialized views were not persisted across restart of the environment.

### CODE REVIEW

File: src/sql-parser/src/ast/defs/statement.rs [hunk 1]
```diff
@@ -1269,6 +1269,18 @@ impl<T: AstInfo> AstDisplay for CreateMaterializedViewStatement<T> {
             f.write_node(cluster);
         }
 
+        if !self.non_null_assertions.is_empty() {
+            f.write_str(" WITH (");
+            for (i, nna) in self.non_null_assertions.iter().enumerate() {
+                f.write_str("ASSERT NOT NULL ");
+                f.write_node(nna);
+                if i + 1 != self.non_null_assertions.len() {
+                    f.write_str(", ")
```
Comment (moulimukherjee) [inline src/sql-parser/src/ast/defs/statement.rs:1278]: nit: you can use `&display::comma_separated`?

Comment (umanwizard) [reply src/sql-parser/src/ast/defs/statement.rs:1278]: yeah, but then I'd have to make a separate node just for `ASSERT NOT NULL <identifier>`,  which seems to require more code than just inlining the comma-separated logic here.

Comment (moulimukherjee) [reply src/sql-parser/src/ast/defs/statement.rs:1278]: cool 👍

**Review by moulimukherjee — APPROVED** [review]
lgtm!

**Review by def- — APPROVED** [review]
Thanks for quickly fixing this.

**Comment by umanwizard** [comment]
Per discussion with @def- , we are merging this with https://github.com/MaterializeInc/materialize/pull/22339 and landing both as one unit. Since https://github.com/MaterializeInc/materialize/pull/22339 has already been reviewed by the QA team, those commits do not need to be reviewed again here.

**Comment by def-** [comment]
Now the data seems wrong, test is still failing in the same scenario:
```
> SELECT * FROM not_null_view1
rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 49s 74s 77s
^^^ +++
31:1: error: non-matching rows: expected:
[["14", "<null>", "16"], ["17", "18", "<null>"], ["24", "<null>", "26"], ["27", "28", "<null>"], ["4", "<null>", "6"], ["7", "8", "<null>"]]
got:
[["4", "<null>", "6"], ["7", "8", "<null>"]]
Poor diff:
- 14 <null> 16
- 17 18 <null>
- 24 <null> 26
- 27 28 <null>

     |
  30 | 
  31 | > SELECT * FROM not_null_view1
     | ^
```

**Comment by umanwizard** [comment]
That is strange .. I'll take a look.

**Comment by umanwizard** [comment]
@def- it seems the `validate` can return different results if run multiple times consecutively, since it deletes some values and then inserts possibly different values. Can't this cause issues in the testing framework?

**Comment by umanwizard** [comment]
I confirmed that the bug is in the test.

This is a minimized script of all the actions taken by the `RestartEnvironmentdClusterdStorage` scenario on the `MaterializedViewsAssertNotNull` check until the point where we fail:

```
CREATE TABLE not_null_table (x INT, y INT, z INT);
INSERT INTO not_null_table VALUES (NULL, 2, 3), (4, NULL, 6), (7, 8, NULL);
CREATE MATERIALIZED VIEW not_null_view1 WITH (ASSERT NOT NULL x) AS SELECT * FROM not_null_table;
CREATE MATERIALIZED VIEW not_null_view2 WITH (ASSERT NOT NULL y) AS SELECT * FROM not_null_table;
INSERT INTO not_null_table VALUES (NULL, 12, 13), (14, NULL, 16), (17, 18, NULL);
CREATE MATERIALIZED VIEW not_null_view3 WITH (ASSERT NOT NULL z) AS SELECT * FROM not_null_table;
INSERT INTO not_null_table VALUES (NULL, 22, 23), (24, NULL, 26), (27, 28, NULL);
DELETE FROM not_null_table WHERE x IS NULL;
DELETE FROM not_null_table WHERE y IS NULL;
DELETE FROM not_null_table WHERE z IS NULL;
INSERT INTO not_null_table VALUES (NULL, 2, 3), (4, NULL, 6), (7, 8, NULL);
-- Note that here we are starting a new `validate` run.
DELETE FROM not_null_table WHERE x IS NULL;
SELECT * FROM not_null_view1;
```

Inspecting this script, the logically expected contents of `not_null_view1` are

```
 x | y | z 
---+---+---
 4 |   | 6
 7 | 8 |  
```

which is indeed what is returned.

**Comment by def-** [comment]
My bad, let me quickly fix the test.

**Pull request merged**